cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/BootstrapService.java

```diff
@@ -0,0 +1,196 @@
+    executorService.shutdownNow();
+    LOG.info("Stopped {}", getClass().getSimpleName());
+  }
+
+  /**
+   * @return whether the system is bootstrapped or not
+   */
+  public boolean isBootstrapped() {
+    return bootstrapStore.isBootstrapped();
+  }
+
+  /**
+   * Execute all steps in the loaded bootstrap config without skipping any of them.
+   *
+   * @return the result of executing the bootstrap steps.
+   * @throws IllegalStateException if bootstrapping is already in progress
+   * @throws InterruptedException if bootstrapping was interrupted
+   */
+  public BootstrapResult bootstrap() throws InterruptedException {
+    return bootstrap(x -> false);
+  }
+
+  /**
+   * Execute the steps in the loaded bootstrap config.
+   *
+   * @param shouldSkip predicate that determines whether to skip a step
+   * @return the result of executing the bootstrap steps.
+   * @throws IllegalStateException if bootstrapping is already in progress
+   * @throws InterruptedException if bootstrapping was interrupted
+   */
+  public BootstrapResult bootstrap(Predicate<BootstrapStep> shouldSkip) throws InterruptedException {
+    List<BootstrapStepResult> results = new ArrayList<>(config.getSteps().size());
+    if (!bootstrapping.compareAndSet(false, true)) {
+      throw new IllegalStateException("Bootstrap already in progress.");
+    }
+
+    try {
+      for (BootstrapStep step : config.getSteps()) {
+        results.add(executeStep(step, shouldSkip));
+      }
+      if (!config.getSteps().isEmpty()) {
+        bootstrapStore.bootstrapped();
+      }
+      return new BootstrapResult(results);
+    } finally {
+      bootstrapping.set(false);
+    }
+  }
+
+  private BootstrapStepResult executeStep(BootstrapStep step,
+                                          Predicate<BootstrapStep> shouldSkip) throws InterruptedException {
+    try {
+      step.validate();
+    } catch (IllegalArgumentException e) {
+      LOG.warn("Bootstrap step {} failed because it is malformed: {}", step.getLabel(), e.getMessage());
+      return new BootstrapStepResult(step.getLabel(), BootstrapStepResult.Status.FAILED, e.getMessage());
+    }
+
+    if (shouldSkip.test(step)) {
+      return new BootstrapStepResult(step.getLabel(), BootstrapStepResult.Status.SKIPPED, null);
+    }
+
+    BootstrapStepExecutor executor = executors.get(step.getType());
+    if (executor == null) {
+      // should not be possible, as deserialization of the file into a BootStrapConfig should have failed
+      return new BootstrapStepResult(step.getLabel(), BootstrapStepResult.Status.FAILED,
+                                     String.format("Unknown bootstrap step type '%s' for '%s'.",
+                                                   step.getType(), step.getLabel()));
+    }
+    return executor.execute(step.getLabel(), step.getArguments());
+  }
+
+  /**
+   * Reloads the bootstrap config.
+   */
+  public void reload() {
```

Comment:
is this only used in test?

Reply:
currently yes, but it will be called when we add the rest endpoint.